iceoryx_utils/source/posix_wrapper/timer.cpp

```diff
@@ -76,52 +76,93 @@ void Timer::OsTimer::callbackHelper(sigval data)
+    // via m_timerInvocationCounter and the other thread will do this for us
+    // note that thte invocation may be lost if the callback is already running, this is intentional
+    // to avoid an unlimited number of callback invocations to pile up
+
+    // note: if we can tolerate those calls to pile up we can get rid of this flag and the while loop as well
+
+    if (!handle.m_callbackIsAboutToBeExecuted.test_and_set(std::memory_order_acq_rel))
     {
-        if (handle.m_accessMutex.try_lock() == false)
+        // flag protected region
+
+        // the mutex is needed to protect against concurrent deletion of the timer in the destructor
+        std::lock_guard<std::mutex> lockGuard(handle.m_accessMutex);
+
+        // clear the flag regardless of how we leave the function
+        cxx::GenericRAII clearCallbackFlag(
+            []() {}, [&]() { handle.m_callbackIsAboutToBeExecuted.clear(std::memory_order_release); });
+
+        do
         {
-            if (handle.m_catchUpPolicy == CatchUpPolicy::TERMINATE)
+            // prohibits other threads from entering the flag protected region
+            handle.m_callbackIsAboutToBeExecuted.test_and_set(std::memory_order_acq_rel);
+
+            if (handle.m_timer == nullptr)
             {
-                errorHandler(Error::kPOSIX_TIMER__CALLBACK_RUNTIME_EXCEEDS_RETRIGGER_TIME);
+                errorHandler(Error::kPOSIX_TIMER__INCONSISTENT_STATE);
+                return;
             }
-            return;
-        }
-        iox::cxx::GenericRAII lockGuard([] {}, [&handle] { handle.m_accessMutex.unlock(); });
 
-        if (handle.m_timer == nullptr)
-        {
-            errorHandler(Error::kPOSIX_TIMER__INCONSISTENT_STATE);
-            return;
-        }
+            if (!handle.m_inUse.load(std::memory_order::memory_order_relaxed))
+            {
+                return;
+            }
 
-        if (!handle.m_inUse.load(std::memory_order::memory_order_relaxed))
-        {
-            return;
-        }
+            if (descriptor != handle.m_descriptor.load(std::memory_order::memory_order_relaxed))
+            {
+                return;
+            }
 
-        if (descriptor != handle.m_descriptor.load(std::memory_order::memory_order_relaxed))
-        {
-            return;
-        }
+            if (!handle.m_isTimerActive.load(std::memory_order::memory_order_relaxed))
+            {
+                return;
+            }
 
-        if (!handle.m_isTimerActive.load(std::memory_order::memory_order_relaxed))
-        {
-            return;
-        }
+            invocationCounter = handle.m_timerInvocationCounter.exchange(0u, std::memory_order_acq_rel);
+
+            // did someone else execute the callback for us?
+            if (invocationCounter != 0u)
+            {
+                handle.m_timer->executeCallback();
+            }
+
+            handle.m_callbackIsAboutToBeExecuted.clear(std::memory_order_release);
```

Comment:
from here up to the point where the flag is set again, there is a window where the CatchUpPolicy::TERMINATE will be handled wrong